shell/src/main/java/org/apache/kafka/shell/MetadataNodeManager.java

```diff
@@ -403,4 +424,149 @@ static List<String> clientQuotaRecordDirectories(List<EntityData> entityData) {
+                break;
+            }
+            case REMOVE_RANGE_RECORD: {
+                RemoveRangeRecord record = (RemoveRangeRecord) message;
+                data.root.rmrf("s3Streams", Long.toString(record.streamId()), "ranges", Integer.toString(record.rangeIndex()));
+                break;
+            }
+            case S3_STREAM_OBJECT_RECORD: {
+                S3StreamObjectRecord record = (S3StreamObjectRecord) message;
+                DirectoryNode s3StreamSetObject = data.root.mkdirs("s3Streams", Long.toString(record.streamId()), "s3StreamObjects", Long.toString(record.objectId()));
+                s3StreamSetObject.create("startOffset").setContents(Long.toString(record.startOffset()));
+                s3StreamSetObject.create("endOffset").setContents(Long.toString(record.endOffset()));
+                s3StreamSetObject.create("dataTimeInMs").setContents(Long.toString(record.dataTimeInMs()));
+                break;
+            }
+            case REMOVE_S3_STREAM_OBJECT_RECORD: {
+                RemoveS3StreamObjectRecord record = (RemoveS3StreamObjectRecord) message;
+                data.root.rmrf("s3Objects", Long.toString(record.objectId()));
+                data.root.rmrf("s3Streams", Long.toString(record.streamId()), "s3StreamObjects", Long.toString(record.objectId()));
+                break;
+            }
+            case S3_STREAM_SET_OBJECT_RECORD: {
+                S3StreamSetObjectRecord record = (S3StreamSetObjectRecord) message;
+                DirectoryNode streamSetObject = data.root.mkdirs("nodes", Integer.toString(record.nodeId()), "s3StreamSetObjects", Long.toString(record.objectId()));
+                streamSetObject.create("orderId").setContents(Long.toString(record.orderId()));
+                streamSetObject.create("dataTimeInMs").setContents(Long.toString(record.dataTimeInMs()));
+                byte[] bytes = record.ranges();
+                // Decode ranges and convert it to human-readable string.
+                List<StreamOffsetRange> ranges = S3StreamSetObject.decode(bytes);
+                streamSetObject.create("ranges").setContents(ranges.toString());
+                break;
+            }
+            case REMOVE_STREAM_SET_OBJECT_RECORD: {
+                RemoveStreamSetObjectRecord record = (RemoveStreamSetObjectRecord) message;
+                data.root.rmrf("s3Objects", Long.toString(record.objectId()));
+                data.root.rmrf("nodes", Integer.toString(record.nodeId()), "s3StreamSetObjects", Long.toString(record.objectId()));
+                break;
+            }
+            case S3_OBJECT_RECORD: {
+                S3ObjectRecord record = (S3ObjectRecord) message;
+                DirectoryNode s3Objects = data.root.mkdirs("s3Objects");
+                DirectoryNode s3Object = s3Objects.mkdirs(Long.toString(record.objectId()));
+                s3Object.create("objectSize").setContents(Long.toString(record.objectSize()));
```

Comment:
Change record to a file

Reply:
The latest hierarchy already uses a single file whenever possible.